keepassxreboot/keepassxc
# Draft: Libvirt auto-type support / auto-type extensibility

I was missing a way to use credentials on headless VMs/guests (e.g. server guests or other types of guests without a remote viewer such as SPICE/VNC), so I decided to extend the auto-type feature with support for [libvirt's `virDomainSendKey`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainSendKey).

While this PR probably as functional as it's going to get, I'm still not really happy with the naming of some stuff like "external plugin". If you can think of a better names, let me know.

#### New plugin interface

Since the current auto-type plugin interface is very specifically built for platform integration and the libvirt feature requires explicit target selection I decided to create a new plugin interface. The libvirt plugin implements this new interface.

#### Target selection

The new plugin interface allows the implementation to (optionally) return a list of available targets. On auto-type the user is presented with a selection dialog to pick the desired target from the list. Upon selection, the plugin implementation gets the selected target provided with the auto-type action. 

#### Limitations

The plugin uses [libvirt's `virDomainSendKey`](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainSendKey) to send keystrokes, and sadly is constrained to a [limited key map](https://github.com/qemu/keycodemapdb/blob/master/data/keymaps.csv). So no unicode unless someone can figure out a reliable way to do that. I took a gander at [hex code input](https://en.wikipedia.org/wiki/Unicode_input#Hexadecimal_input), but that doesn't seem a great solution. 

#### Dead keys

In my experience the default Windows keyboard layout is US International (with dead keys), so I added some basic dead keys support. Dead keys are supported in the sense that it will properly output stuff like double quotes, backticks, and carets (but no accented letters). Since dead keys can be enabled/disabled in the OS, I added two new settings. One allows to enable/disable dead keys for Windows guests (best effort guest OS detection), the other allows to enable/disable dead keys for other guest operating systems.

## Screenshots

![context-menu](https://user-images.githubusercontent.com/2717179/112392930-8837c100-8cfa-11eb-8fb1-84ee0b285b9f.png)
![target-select](https://user-images.githubusercontent.com/2717179/112392947-8c63de80-8cfa-11eb-9ea0-7f4ac7d2bbdd.png)
![dead-keys-settings](https://user-images.githubusercontent.com/2717179/112393357-38a5c500-8cfb-11eb-8687-f9877e253341.png)

## Testing strategy

Functional:

- Create one or more libvirt domains (easiest to do with `virt-manager`)
- Make sure one of the libvirt domains is running (and shows up in `virsh list`)
- Open context menu "Perform Auto-Type To Plugin" -> "Libvirt"
- Select target libvirt domain
- ???
- Profit

Added two basic unit tests for the new plugin interface. Not sure what/how to automatically test other stuff. 

## To do

- [ ] Figure out why persistence of the two new settings seems broken. Could've sworn it worked at some point... :sweat_smile: 

## Type of change

- ✅ New feature (change that adds functionality)
